webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts

```diff
@@ -0,0 +1,203 @@
+ * the next call to queueForLoading will start a new one.
+ *
+ * DelayedDataLoader is intended to be used for loading data for components which are unaware of each other and may appear
+ * in different places in the UI from each other which could otherwise send repeated requests for the same or similar
+ * data as one another.
+ */
+export class DelayedDataLoader<Identifier> extends DataLoader<Identifier, Promise<unknown>> {
+    private readonly wait: number = -1;
+
+    private timeoutId: number = -1;
+    private timeoutCallbacks = new Set<{
+        identifiers: Set<Identifier>;
+        resolve(): void;
+    }>();
+
+    constructor(args: {
+        fetchBatch: (identifiers: Identifier[]) => Promise<unknown>;
+        maxBatchSize: number;
+        wait: number;
+    }) {
+        super(args);
+
+        this.wait = args.wait;
+    }
+
+    public queueForLoading(identifiersToLoad: Identifier[]): void {
+        super.queueForLoading(identifiersToLoad);
+
+        this.startTimeoutIfNeeded();
+    }
+
+    public queueAndWait(identifiersToLoad: Identifier[]): Promise<void> {
+        return new Promise((resolve) => {
+            super.queueForLoading(identifiersToLoad);
+
+            // Save the callback that will resolve this promise so that the caller of this method can wait for its
+            // data to be loaded
+            this.timeoutCallbacks.add({
+                identifiers: new Set(identifiersToLoad),
+                resolve,
+            });
+
+            this.startTimeoutIfNeeded();
+        });
+    }
+
+    private startTimeoutIfNeeded(): void {
+        if (this.timeoutId !== -1) {
+            return;
+        }
+
+        this.timeoutId = window.setTimeout(() => {
+            // Ensure that timeoutId and timeoutCallbacks are cleared before doing anything async so that any calls to
+            // queueForLoading which come while we're fetching this batch are added to the next batch instead
+            this.timeoutId = -1;
+
+            const {identifiers, moreToLoad} = this.startBatch();
+
+            // Start another timeout if there's still more data to load
+            if (moreToLoad) {
+                this.startTimeoutIfNeeded();
+            }
+
+            this.fetchBatch(identifiers).then(() => this.resolveCompletedCallbacks(identifiers));
+        }, this.wait);
+    }
+
+    private resolveCompletedCallbacks(identifiers: Identifier[]): void {
+        for (const callback of this.timeoutCallbacks) {
+            for (const identifier of identifiers) {
+                callback.identifiers.delete(identifier);
+            }
+
+            if (callback.identifiers.size === 0) {
+                this.timeoutCallbacks.delete(callback);
+                callback.resolve();
+            }
+        }
+    }
+
+    public isBusy(): boolean {
+        return super.isBusy() || this.timeoutCallbacks.size > 0 || this.timeoutId !== -1;
+    }
```

Comment:
Yeah, I can do that. I wish there was an easier way to expose stuff only for testing, but I can make it protected and extend the class in the test to make it public. I wish JS had the "default" visibility that Java has since it would allow for this behaviour

Reply:
I'm actually going to leave this as-is. Ideally, we wouldn't expose it, but because there's no easy way to only expose it for tests, I don't think it's worth the mental overhead to hide an arguably private method that'll be useless for whoever uses this